org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/textmate/TextMateGrammar.java

```diff
@@ -173,15 +176,54 @@ protected TextMateRule getKeywordControlRule(Grammar grammar, boolean ignoreCase
 		matchBuilder.append("\\b(");
 		List<String> allKeywords = GrammarUtil.getAllKeywords(grammar)
 				.stream()
-				.filter(s->s.matches("\\w+"))
+				.filter(s -> s.matches("\\w+"))
 				.sorted(Comparator.naturalOrder())
 				.collect(Collectors.toList());
-		matchBuilder.append(Joiner.on("|").join(allKeywords));
+		Joiner.on("|").appendTo(matchBuilder, allKeywords);
 		matchBuilder.append(")\\b");
 		MatchRule result = new MatchRule();
 		result.setName("keyword.control." + getLanguageName(grammar));
 		result.setMatch(matchBuilder.toString());
 		return result;
 	}
 
+	protected TextMateRule getPunctuationRule(Grammar grammar, boolean ignoreCase) {
+		StringBuilder matchBuilder = new StringBuilder();
+		if (ignoreCase) {
+			matchBuilder.append("(?i)");
+		}
+		matchBuilder.append("(");
+		List<String> allKeywords = GrammarUtil.getAllKeywords(grammar)
```

Comment:
It seems as if we could use the same implementation for bother regex rules. The only meaningful difference seems to be the `s.matches("\\w+")` vs its negation and the name of the match rule.
Regular keywords should still work if we funnel them through `escapeAndAddWordBoundaries`

Would that work?

Reply:
That works and I've updated the PR accordingly. Only downside is that the generated regex for the `keyword.control` rule is less optimized: `(\\ba\\b|\\bb\\b|\\bc\\b)` instead of `\\b(a|b|c)\\b`